engine/src/main/java/org/terasology/engine/core/module/ModuleManager.java

```diff
@@ -105,49 +98,61 @@ public ModuleManager(String masterServerAddress, List<Class<?>> classesOnClasspa
+        engineDep.setMaxVersion(engineModule.getVersion().getNextMinorVersion());
+
+        HashSet<Name> engineModules = Sets.newHashSet(engineModule.getId());
+        engineModules.addAll(engineModule.getMetadata().getDependencies().stream().map(DependencyInfo::getId).collect(Collectors.toList()));
 
-        registry.stream().filter(mod -> mod != engineModule).forEach(mod -> mod.getMetadata().getDependencies().add(engineDep));
+        registry.stream()
+                .filter(mod -> !engineModules.contains(mod.getId()))
+                .forEach(mod -> mod.getMetadata().getDependencies().add(engineDep));
     }
 
     private void loadModulesFromApplicationPath(PathManager pathManager) {
-        ModulePathScanner scanner = new ModulePathScanner(new ModuleLoader(metadataReader));
-        scanner.getModuleLoader().setModuleInfoPath(TerasologyConstants.MODULE_INFO_FILENAME);
-        scanner.scan(registry, pathManager.getModulePaths());
+        ModulePathScanner scanner = new ModulePathScanner(moduleFactory);
+        List<File> paths = pathManager.getModulePaths().stream().map(Path::toFile).collect(Collectors.toList());
+        scanner.scan(registry, paths);
     }
 
     private Module loadEngineModule(List<Class<?>> classesOnClasspathsToAddToEngine) {
-        Module engineModule;
-        try (Reader reader = new InputStreamReader(getClass().getResourceAsStream("/engine-module.txt"), TerasologyConstants.CHARSET)) {
-            ModuleMetadata metadata = metadataReader.read(reader);
-            List<Class<?>> additionalClassesList = new ArrayList<>(classesOnClasspathsToAddToEngine.size() + 2);
-            additionalClassesList.add(Module.class); // provide access to gestalt-module.jar
-            additionalClassesList.add(Asset.class); // provide access to gestalt-asset-core.jar
-            additionalClassesList.add(UIWidget.class); // provide access to nui.jar
-            additionalClassesList.add(TypeRegistry.class); // provide access to nui-reflect.jar
-            additionalClassesList.add(KeyboardDevice.class); // provide access to nui-input.jar
-            additionalClassesList.add(ModuleTypeRegistry.class); // provide access to nui-gestalt7.jar
-            additionalClassesList.addAll(classesOnClasspathsToAddToEngine); // provide access to any facade-provided classes
-            Class<?>[] additionalClassesArray = new Class[additionalClassesList.size()];
-            additionalClassesArray = additionalClassesList.toArray(additionalClassesArray);
-            engineModule = ClasspathModule.create(metadata, getClass(), additionalClassesArray);
-        } catch (IOException e) {
-            throw new RuntimeException("Failed to read engine metadata", e);
-        } catch (URISyntaxException e) {
-            throw new RuntimeException("Failed to convert engine library location to path", e);
-        }
+        Module packageModule = moduleFactory.createPackageModule("org.terasology.engine");
```

Comment:
What is a "package module"? Does it just collect everything that is within the package? Can other modules add to this engine module package?

Reply:
> Does it just collect everything that is within the package?

Yep, https://github.com/MovingBlocks/gestalt/blob/f459dac2fd0b073f5a76124b106aa491d5ddc2ba/gestalt-module/src/main/java/org/terasology/gestalt/module/ModuleFactory.java#L268-L274

> Can other modules add to this engine module package?

Worrying question, right? Because there's nothing stopping any jar from containing a class under `org/terasology/engine`.

I think/hope that since module jars aren't loaded from the classpath with the main application classloader, there should/could be something about the ModuleClassLoader that prevents them from being conflated with the package module? But there is an awful lot of "should" and "hope" in that statement.

Reply:
Alright, at least that confirms both my assumption and suspicion. 

Do we have a diagram of class and module loaders, class paths, and all that somewhere?